File: cDate.py
import calendar, datetime, math, re;

gbDebugOutput = False;

rDate = re.compile(
  r"^\s*" +
  r"(\d{4})" +
  r"[\-\/]" +
  r"(\d{1,2})" +
  r"[\-\/]" + 
  r"(\d{1,2})" +
  r"\s*$"
);
asMonths = [
  "January", "February", "March", "April", "May", "June",
  "July", "August", "September", "October", "November", "December"
];
asOrdinalPostfixes = [
  "st", "nd", "rd", "th", "th", "th", "th", "th", "th", "th", # 1-10
  "th", "th", "th", "th", "th", "th", "th", "th", "th", "th", # 11-20
  "st", "nd", "rd", "th", "th", "th", "th", "th", "th", "th", # 21-30
  "st"                                                        # 31
];
def fbIsValidInteger(uValue, uMinValueInclusive = None, uMaxValueExclusive = None):
  return (
    isinstance(uValue, int)
    and (uValue % 1 == 0)
    and (uValue >= uMinValueInclusive if uMinValueInclusive is not None else True)
    and (uValue < uMaxValueExclusive if uMaxValueExclusive is not None else True)
  );

class cDate(object):
  # Static methods
  @staticmethod
  def fbIsValidYear(uValue):
    return fbIsValidInteger(uValue, 0);
  @staticmethod
  def fbIsValidMonth(uValue):
    return fbIsValidInteger(uValue, 1, 13);
  @staticmethod
  def fbIsValidDay(uValue):
    return fbIsValidInteger(uValue, 1, 32); # We do not know the month but 32 is never valid.
  @staticmethod
  def fbIsValidDate(uYear, uMonth, uDay):
    return (
      cDate.fbIsValidYear(uYear)
      and cDate.fbIsValidMonth(uMonth)
      and fbIsValidInteger(uDay, 1, cDate.fuGetLastDayInMonth(uYear, uMonth) + 1)
    );
  @staticmethod
  def fsGetDateString(uYear, uMonth, uDay):
    return "%04d-%02d-%02d" % (uYear, uMonth, uDay);
  @staticmethod
  def fuGetLastDayInMonth(uYear, uMonth):
    return calendar.monthrange(uYear, uMonth)[1];
  
  @staticmethod
  def fo0FromPyDate(oDate):
    return None if oDate is None else cDate.foFromPyDate(oDate);
  @staticmethod
  def foFromPyDate(oDate):
    return cDate(oDate.year, oDate.month, oDate.day);
  
  @staticmethod
  def fo0FromJSON(s0Date):
    return None if s0Date is None else cDate.foFromJSON(s0Date);
  @staticmethod
  def foFromJSON(sDate):
    # JSON encoding uses the "string value" of cDate.
    return cDate.foFromString(sDate);
  
  @staticmethod
  def fo0FromMySQL(s0Date):
    return None if s0Date is None else cDate.foFromMySQL(s0Date);
  @staticmethod
  def foFromMySQL(sDate):
    # MySQL encoding uses the "string value" of cDate.
    return cDate.foFromMySQL(sDate);
  @staticmethod
  def fo0FromMySQLDateTime(s0DateTime):
    return None if s0DateTime is None else cDate.foFromMySQLDateTime(s0DateTime);
  @staticmethod
  def foFromMySQLDateTime(sDateTime):
    # MySQL format is "YYYY-MM-DD hh:mm:ss", so we can just split it at the space and use the first part:
    return cDate.foFromMySQL(sDateTime.split(" ")[0]);
  
  @staticmethod
  def fbIsValidDateString(sDate):
    return isinstance(sDate, str) and rDate.match(sDate) is not None;
  @staticmethod
  def fo0FromString(s0Date):
    return None if s0Date is None else cDate.foFromString(s0Date);
  @staticmethod
  def foFromString(sDate):
    oDateMatch = rDate.match(sDate) if isinstance(sDate, str) else None;
    if oDateMatch is None: raise ValueError("Invalid date string " + repr(sDate) + ".");
    return cDate(int(oDateMatch.group(1)), int(oDateMatch.group(2)), int(oDateMatch.group(3)));
  
  @staticmethod
  def foNow():
    return cDate.foFromPyDate(datetime.datetime.now());
  @staticmethod
  def foNowUTC():
    return cDate.foFromPyDate(datetime.datetime.utcnow());
  
  # Constructor
  def __init__(oSelf, uYear, uMonth, uDay):
    if not cDate.fbIsValidDate(uYear, uMonth, uDay): raise ValueError("Invalid date (%s, %s, %s)." % (repr(uYear), repr(uMonth), uDay));
    oSelf.__uYear = uYear;
    oSelf.__uMonth = uMonth; # 1 = January
    oSelf.__uDay = uDay; # 1 = first day of month
  # Properties
  @property
  def uYear(oSelf):
    return oSelf.__uYear;
  @uYear.setter
  def uYear(oSelf, uYear):
    if not cDate.fbIsValidYear(uYear): raise ValueError("Invalid year " + repr(uYear) + ".");
    if not cDate.fbIsValidDate(uYear, oSelf.__uMonth, oSelf.__uDay): raise ValueError("Invalid year in date %s." % cDate.fsGetDateString(uYear, oSelf.__uMonth, oSelf.__uDay));
    oSelf.__uYear = uYear;
  @property
  def uMonth(oSelf):
    return oSelf.__uMonth;
  @uMonth.setter
  def uMonth(oSelf, uMonth):
    if not cDate.fbIsValidMonth(uMonth): raise ValueError("Invalid month " + repr(uMonth) + ".");
    if not cDate.fbIsValidDate(oSelf.__uYear, uMonth, oSelf.__uDay): raise ValueError("Invalid month in date %s." % cDate.fsGetDateString(oSelf.__uYear, uMonth, oSelf.__uDay));
    oSelf.__uMonth = uMonth;
  @property
  def uDay(oSelf):
    return oSelf.__uDay;
  @uDay.setter
  def uDay(oSelf, uDay):
    if not cDate.fbIsValidDay(uDay): raise ValueError("Invalid day " + repr(uDay) + ".");
    if not cDate.fbIsValidDate(oSelf.__uYear, oSelf.__uMonth, uDay): raise ValueError("Invalid day in date %s." % cDate.fsGetDateString(uYear, uMonth, uDay));
    oSelf.__uDay = uDay;
  #methods
  def foClone(oSelf):
    return cDate(oSelf.__uYear, oSelf.__uMonth, oSelf.__uDay);
  
  def fSet(oSelf, uYear, uMonth, uDay):
    if not cDate.fbIsValidDate(uYear, uMonth, uDay): raise ValueError("Invalid date (%s, %s, %s)." % (repr(uYear), repr(uMonth), uDay));
    oSelf.__uYear = uYear;
    oSelf.__uMonth = uMonth;
    oSelf.__uDay = uDay;
  
  def foGetEndDateForDuration(oSelf, oDuration):
    # Note that this code ignores the time (if any) in oDuration
    # Add the year and month:
    iNewYear = oSelf.__uYear + oDuration.iYears;
    iNewMonth0Based = oSelf.__uMonth - 1 + oDuration.iMonths;
    if gbDebugOutput: print("year %s %s %s => %s, month (base 0) %s %s %s => %s" % (
      repr(oSelf.__uYear), "-" if oDuration.iYears < 0 else "+", abs(oDuration.iYears), repr(iNewYear),
      repr(oSelf.__uMonth - 1), "-" if oDuration.iMonths < 0 else "+", abs(oDuration.iMonths), repr(iNewMonth0Based),
    ));
    # If uNewMonth < 0 or > 11, convert the excess/shortage to years and add it.
    iMonthsExcessOrShortageInYears = math.floor(iNewMonth0Based / 12);
    iNewYear += iMonthsExcessOrShortageInYears;
    assert iNewYear > 0, \
        "Year cannot be < 0 (%s)" % iNewYear;
    uNewYear = iNewYear;
    uNewMonth0Based = iNewMonth0Based % 12;
    uNewMonth = uNewMonth0Based + 1;
    if gbDebugOutput: print("year %s= %s => %s, month (base 0) %s => %s => base 1: %s" % (
      "-" if iMonthsExcessOrShortageInYears < 0 else "+", abs(iMonthsExcessOrShortageInYears), repr(uNewYear),
      repr(iNewMonth0Based), repr(uNewMonth0Based), repr(uNewMonth),
    ));
    # If we added months and ended up in another month in which the current day does not exist (e.g. Feb 31st)
    # reduce the day (i.e. Feb 28th/29th)
    uLastDayInNewMonth = cDate.fuGetLastDayInMonth(uNewYear, uNewMonth);
    uNewDayInNewMonth = oSelf.__uDay if oSelf.uDay <= uLastDayInNewMonth else uLastDayInNewMonth;
    if gbDebugOutput and uNewDayInNewMonth != oSelf.__uDay: print("day %d => %d" % (
      oSelf.__uDay, uNewDayInNewMonth
    ));
    # Add the days by creating the Python datetime.date equivalent and adding the days using datetime.timedelta, then
    # converting back to cDate. This allows us to reuse the Python API for tracking the number of days in each month.
    oEndDate = cDate.foFromPyDate(
      datetime.date(uNewYear, uNewMonth, uNewDayInNewMonth)
      + datetime.timedelta(oDuration.iDays)
    );
    return oEndDate;
  
  def foGetDurationForEndDate(oSelf, oEndDate):
    if gbDebugOutput: print("=== cDate.foGetDurationForEndDate(%s, %s) ===" % (str(oSelf), str(oEndDate)));
    # If the end date is before this date, the duration is going to be negative.
    # To keep this code simple to review, we always calculate the positive duration between the two dates
    # and later invert it should the real result be a negative duration.
    bNegativeDuration = oEndDate.fbIsBefore(oSelf);
    uDurationMultiplier = -1 if bNegativeDuration else 1; # Used to potentially invert the duration later.
    oFirstDate = oEndDate if bNegativeDuration else oSelf;
    oLastDate = oSelf if bNegativeDuration else oEndDate;
    uDurationYears = oLastDate.uYear - oFirstDate.uYear;
    uDurationMonths = oLastDate.uMonth - oFirstDate.uMonth;
    uDurationDays = oLastDate.uDay - oFirstDate.uDay;
    if gbDebugOutput: print("  1: %+dy%+dm%+dd*%+d" % (uDurationYears, uDurationMonths, uDurationDays, uDurationMultiplier));
    # The number of days in the last month various
    uDaysInLastDatesPreviousMonth = cDate.fuGetLastDayInMonth(oLastDate.uYear - (1 if oLastDate.uMonth == 1 else 0), ((oLastDate.uMonth + 10) % 12) + 1);
    if uDurationDays >= oLastDate.uDay:
      # If uDurationDays > last date's day, adding the days moved it into a new month; convert this into a month and adjust the days.
      # e.g. 2000-1-31 -> 2000-2-2 => -1m+29d (at this point) => +2d (after this adjustment)
      uDurationMonths += 1;
      uDurationDays = uDaysInLastDatesPreviousMonth - uDurationDays;
      if gbDebugOutput: print("  2: d->m: %+dy%+dm%+dd*%+d" % (uDurationYears, uDurationMonths, uDurationDays, uDurationMultiplier));
    elif uDurationDays < 0:
      # If uDurationDays < 0, the day is before adding the days moved it into a new month; convert this into a month and adjust the days.
      # e.g. 2000-1-2 -> 2000-2-1 => +1m-1d (at this point) => +30d (after this adjustment)
      uDurationMonths -= 1;
      uDurationDays += uDaysInLastDatesPreviousMonth;
      if gbDebugOutput: print("  2: d->m: %+dy%+dm%+dd*%+d" % (uDurationYears, uDurationMonths, uDurationDays, uDurationMultiplier));
    # If uDurationMonths < 0 or >= 12, convert the excess to years and add them.
    if uDurationMonths < 0 or uDurationMonths >= 12:
      uDurationYears += math.floor(uDurationMonths / 12);
      uDurationMonths = uDurationMonths % 12;
      if gbDebugOutput: print("  3: m->y: %+dy%+dm%+dd*%+d" % (uDurationYears, uDurationMonths, uDurationDays, uDurationMultiplier));
    from .cDateDuration import cDateDuration;
    oDuration = cDateDuration(
      uDurationYears * uDurationMultiplier,
      uDurationMonths * uDurationMultiplier,
      uDurationDays * uDurationMultiplier,
    );
    if gbDebugOutput: print("=> return %s" % oDuration);
    return oDuration;
  
  def fbIsBefore(oSelf, oDate):
    if oSelf.__uYear < oDate.uYear: return True;
    if oSelf.__uYear > oDate.uYear: return False;
    if oSelf.__uMonth < oDate.uMonth: return True;
    if oSelf.__uMonth > oDate.uMonth: return False;
    if oSelf.__uDay < oDate.uDay: return True;
    #if oSelf.__uDay > oDate.uDay: return False;
    return False;
  def fbIsEqualTo(oSelf, oDate):
    return oSelf.__uYear == oDate.uYear and oSelf.__uMonth == oDate.uMonth and oSelf.__uDay == oDate.uDay;
  def fbIsAfter(oSelf, oDate):
    if oSelf.__uYear > oDate.uYear: return True;
    if oSelf.__uYear < oDate.uYear: return False;
    if oSelf.__uMonth > oDate.uMonth: return True;
    if oSelf.__uMonth < oDate.uMonth: return False;
    if oSelf.__uDay > oDate.uDay: return True;
    #if oSelf.__uDay < oDate.uDay: return False;
    return False;
  
  def fbIsInThePast(oSelf):
    return cDate.fbIsBefore(oSelf, cDate.foNow());
  def fbIsInThePastUTC(oSelf):
    return cDate.fbIsBefore(oSelf, cDate.foNowUTC());
  def fbIsToday(oSelf):
    return cDate.fbIsEqualTo(oSelf, cDate.foNow());
  def fbIsTodayUTC(oSelf):
    return cDate.fbIsEqualTo(oSelf, cDate.foNowUTC());
  def fbIsInTheFuture(oSelf):
    return cDate.fbIsAfter(oSelf, cDate.foNow());
  def fbIsInTheFutureUTC(oSelf):
    return cDate.fbIsAfter(oSelf, cDate.foNowUTC());
  
  def fsToHumanReadableString(oSelf):
    # Month <day>th, <year>
    return "%s %d%s, %d" % (
      asMonths[oSelf.__uMonth - 1],
      oSelf.__uDay, asOrdinalPostfixes[oSelf.__uDay],
      oSelf.__uYear,
    );
  def foToPyDate(oSelf):
    return datetime.date(oSelf.__uYear, oSelf.__uMonth, oSelf.__uDay);
  def fnToTimestamp(oSelf):
    return time.mktime(cDate.foToPyDate(oSelf).timetuple()) + (oSelf.uMicrosecond / 1000.0 / 1000);
  def fxToJSON(oSelf):
    # JSON encoding uses the "string value" of cDate.
    return cDate.fsToString(oSelf);
  def fsToMySQL(oSelf):
    # MySQL encoding uses the "string value" of cDate.
    return cDate.fsToString(oSelf);
  def fsToString(oSelf):
    return cDate.fsGetDateString(oSelf.__uYear, oSelf.__uMonth, oSelf.__uDay);
  def __str__(oSelf):
    return cDate.fsToString(oSelf);
  
  def __cmp__(oSelf, oOther):
    assert isinstance(oOther, cDate), \
        "Cannot compare %s to %s" % (oSelf, oOther);
    if oSelf.uYear != oSelf.uYear: return oSelf.uYear - oOther.uYear;
    if oSelf.uMonth != oSelf.uMonth: return oSelf.uMonth - oOther.uMonth;
    if oSelf.uDay != oSelf.uDay: return oSelf.uDay - oOther.uDay;
    return 0;

File: cDate.js
"use strict";
const cDate = (function() {
  const rDate = new RegExp(
          "^\\s*" +
          "(\\d{4})" + "[\\-\\/]" +
          "(\\d{1,2})" + "[\\-\\/]" + 
          "(\\d{1,2})" +
          "\\s*$"
        ),
        asMonths = [
          "January", "February", "March", "April", "May", "June",
          "July", "August", "September", "October", "November", "December"
        ],
        asOrdinalPostfixes = [
          "st", "nd", "rd", "th", "th", "th", "th", "th", "th", "th", # 1-10
          "th", "th", "th", "th", "th", "th", "th", "th", "th", "th", # 11-20
          "st", "nd", "rd", "th", "th", "th", "th", "th", "th", "th", # 21-30
          "st"                                                        # 31
        ];
  function fbIsValidYear(xYear) {
    return typeof(xYear) == "number" && !isNaN(xYear);
  };
  function fbIsValidMonth(xMonth) {
    return typeof(xMonth) == "number" && !isNaN(xMonth) && xMonth >= 1 && xMonth <= 12;
  };
  function fbIsValidDay(xDay) {
    return typeof(xDay) == "number" && !isNaN(xDay) && xDay >= 1 && xDay <= 31;
  };
  function fbIsValidDate(uYear, uMonth, uDay) {
    return uDay <= new Date(uYear, uMonth, 0).getDate();
  };
  function fsGetDateString(uYear, uMonth, uDay) {
    return [
      uYear.toString().padStart(4, "0"), 
      uMonth.toString().padStart(2, "0"), 
      uDay.toString().padStart(2, "0"),
    ].join("-");
  };
  function fuGetLastDayInMonth0Based(uYear, uMonth0Based) {
    return new Date(uYear, uMonth0Based + 1, 0).getDate();
  };
  
  // constructor
  function cDate(uYear, uMonth, uDay) {
    if (!this) return new cDate(uYear, uMonth, uDay);
    if (!fbIsValidYear(uYear)) throw new Error("Invalid year " + JSON.stringify(uYear) + ".");
    if (!fbIsValidMonth(uMonth)) throw new Error("Invalid month " + JSON.stringify(uMonth) + ".");
    if (!fbIsValidDay(uDay)) throw new Error("Invalid day " + JSON.stringify(uDay) + ".");
    if (!fbIsValidDate(uYear, uMonth, uDay)) throw new Error("Invalid date " + fsGetDateString(uYear, uMonth, uDay) + ".");
    this.__uYear = uYear;
    this.__uMonth = uMonth; // 1 = January
    this.__uDay = uDay; // 1 = first day of month
  };
  // properties
  Object.defineProperty(cDate.prototype, "uYear", {
    "get": function() { return this.__uYear; },
    "set": function(uYear) {
      if (!fbIsValidYear(uYear)) throw new Error("Invalid year " + JSON.stringify(uYear) + ".");
      if (!fbIsValidDate(uYear, this.__uMonth, this.__uDay)) throw new Error("Invalid year in date " + fsGetDateString(uYear, this.__uMonth, this.__uDay) + ".");
      this.__uYear = uYear;
    },
  });
  Object.defineProperty(cDate.prototype, "uMonth", {
    "get": function() { return this.__uMonth; },
    "set": function(uMonth) {
      if (!fbIsValidMonth(uMonth)) throw new Error("Invalid month " + JSON.stringify(uMonth) + ".");
      if (!fbIsValidDate(this.__uYear, uMonth, this.__uDay)) throw new Error("Invalid month in date " + fsGetDateString(this.__uYear, uMonth, this.__uDay) + ".");
      this.__uMonth = uMonth;
    },
  });
  Object.defineProperty(cDate.prototype, "uDay", {
    "get": function() { return this.__uDay; },
    "set": function(uDay) {
      if (!fbIsValidDay(uDay)) throw new Error("Invalid day " + JSON.stringify(uDay) + ".");
      if (!fbIsValidDate(this.__uYear, this.__uMonth, uDay)) throw new Error("Invalid day in date " + fsGetDateString(this.__uYear, this.__uMonth, uDay) + ".");
      this.__uDay = uDay;
    },
  });
  // static methods
  cDate.fo0FromJSDate = function cDate_fo0FromJSDate(oDate) {
    return oDate === null ? null : cDate.foFromJSDate(oDate);
  };
  cDate.foFromJSDate = function cDate_foFromJSDate(oDate) {
    return new cDate(oDate.getFullYear(), oDate.getMonth() + 1, oDate.getDate());
  };
  cDate.fo0FromJSON = function cDate_fo0FromJSON(s0Date) {
    return s0Date === null ? null : cDate.foFromJSON(s0Date);
  };
  cDate.foFromJSON = function cDate_foFromJSON(sDate) {
    return cDate.foFromString(sDate);
  };
  cDate.fbIsValidDateString = function cDate_fbIsValidDateString(sDate) {
    return typeof(sDate) === "string" && sDate.match(rDate) !== null;
  };
  cDate.fo0FromString = function cDate_fo0FromString(s0Date) {
    return s0Date === null ? null : cDate.foFromString(s0Date);
  };
  cDate.foFromString = function cDate_foFromString(sDate) {
    const oDateMatch = typeof(sDate) === "string" ? sDate.match(rDate) : null;
    if (oDateMatch === null) throw new Error("Invalid date string " + JSON.stringify(sDate) + ".");
    return new cDate(parseInt(oDateMatch[1]), parseInt(oDateMatch[2]), parseInt(oDateMatch[3]));
  };
  cDate.foNow = function cDate_foNow() {
    const oNow = new Date();
    return new cDate(oNow.getFullYear(), oNow.getMonth() + 1, oNow.getDate());
  };
  cDate.foNowUTC = function cDate_foNowUTC() {
    const oNow = new Date();
    return new cDate(oNow.getUTCFullYear(), oNow.getUTCMonth() + 1, oNow.getUTCDate());
  };
  // methods
  cDate.prototype.foClone = function cDate_foClone() {
    return new this.constructor(this.__uYear, this.__uMonth, this.__uDay);
  };
  cDate.prototype.fSet = function cDate_fSet(uYear, uMonth, uDay) {
    if (uYear === undefined || uYear === null) uYear = this.__uYear;
    else if (!fbIsValidYear(uYear)) throw new Error("Invalid year " + JSON.stringify(uYear) + ".");
    if (uMonth === undefined || uMonth === null) uMonth = this.__uMonth;
    else if (!fbIsValidMonth(uMonth)) throw new Error("Invalid month " + JSON.stringify(uMonth) + ".");
    if (uDay === undefined || uDay === null) uDay = this.__uDay;
    else if (!fbIsValidDay(uDay)) throw new Error("Invalid day " + JSON.stringify(uDay) + ".");
    if (!fbIsValidDate(uYear, this.__uMonth, this.__uDay)) throw new Error("Invalid date " + fsGetDateString(uYear, this.__uMonth, this.__uDay) + ".");
    this.__uYear = uYear;
    this.__uMonth = uMonth;
    this.__uDay = uDay;
  };
  cDate.prototype.foGetEndDateForDuration = function cDate_foEndDateForDuration(oDuration) {
    // Note that this code ignores the time (if any) in oDuration
    // Add the year and month:
    let uNewYear = this.__uYear + oDuration.iYears;
    let uNewMonth0Based = this.__uMonth - 1 + oDuration.iMonths;
    // If uNewMonth < 0 or > 11, convert the excess to years and add it.
    uNewYear += Math.floor(uNewMonth0Based / 12);
    uNewMonth0Based = ((uNewMonth0Based % 12) + (uNewMonth0Based < 0 ? 12 : 0)) % 12;
    // If we added months and ended up in another month in which the current day does not exist (e.g. Feb 31st)
    // reduce the day (i.e. Feb 28th/29th)
    const uLastDayInNewMonth = fuGetLastDayInMonth0Based(uNewYear, uNewMonth0Based),
          uNewDayInNewMonth = this.uDay <= uLastDayInNewMonth ? this.__uDay : uLastDayInNewMonth;
    // Add the days by creating the Python datetime.date equivalent and adding the days using datetime.timedelta, then
    // converting back to cDate. This allows us to reuse the Python API for tracking the number of days in each month.
    const oEndDate = cDate.foFromJSDate(
      new Date(uNewYear, uNewMonth0Based, uNewDayInNewMonth + oDuration.iDays)
    );
    return oEndDate;
  };
  cDate.prototype.foGetDurationForEndDate = function cDate_foGetDurationForEndDate(oEndDate) {
    // If the end date is before this date, the duration is going to be negative.
    // To keep this code simple to review, we always calculate the positive duration between the two dates
    // and later invert it should the real result be a negative duration.
    const bNegativeDuration = oEndDate.fbIsBefore(this),
          uDurationMultiplier = bNegativeDuration ? -1 : 1, // Used to potentially invert the duration later.
          oFirstDate = bNegativeDuration ? oEndDate : this,
          oLastDate = bNegativeDuration ? this : oEndDate;
    let uDurationYears = oLastDate.uYear - oFirstDate.uYear,
        uDurationMonths = oLastDate.uMonth - oFirstDate.uMonth,
        uDurationDays = oLastDate.uDay - oFirstDate.uDay,
        // The number of days in the last month various
        uDaysInLastDatesPreviousMonth = new Date(oLastDate.uYear - (oLastDate.uMonth == 1 ? 1 : 0), ((oLastDate.uMonth + 10) % 12 + 1), 0).getDate();
    if (uDurationDays >= oLastDate.uDay) {
      // If uDurationDays > last date's day, adding the days moved it into a new month; convert this into a month and adjust the days.
      // e.g. 2000-1-31 -> 2000-2-2 => -1m+29d (at this point) => +2d (after this adjustment)
      uDurationMonths += 1;
      uDurationDays = uDaysInLastDatesPreviousMonth - uDurationDays;
    } else if (uDurationDays < 0) {
      // If uDurationDays < 0, the day is before adding the days moved it into a new month; convert this into a month and adjust the days.
      // e.g. 2000-1-2 -> 2000-2-1 => +1m-1d (at this point) => +30d (after this adjustment)
      uDurationMonths -= 1;
      uDurationDays += uDaysInLastDatesPreviousMonth;
    };
    // If uDurationMonths < 0 or >= 12, convert the excess to years and add them.
    if (uDurationMonths < 0 || uDurationMonths >= 12) {
      uDurationYears += Math.floor(uDurationMonths / 12) + (uDurationMonths < 0 ? -1 : 0);
      uDurationMonths = (uDurationMonths % 12) + (uDurationMonths < 0 ? 12 : 0);
    };
    const oDuration = new cDateDuration(
            uDurationYears * uDurationMultiplier,
            uDurationMonths * uDurationMultiplier,
            uDurationDays * uDurationMultiplier,
          );
    return oDuration;
  };

  // This object has valueOf(), which returns the number of milliseconds since the epoch, so you can also
  // use (this.valueOf() < oDate.valueOf(), this.valueOf() == oDate.valueOf(), and this.valueOf() > oDate.valueOf())
  cDate.prototype.fbIsBefore = function cDate_fbIsBefore(oDate) {
    if (this.__uYear < oDate.uYear) return true;
    if (this.__uYear > oDate.uYear) return false;
    if (this.__uMonth < oDate.uMonth) return true;
    if (this.__uMonth > oDate.uMonth) return false;
    if (this.__uDay < oDate.uDay) return true;
    //if (this.__uDay > oDate.uDay) return false;
    return false;
  };
  cDate.prototype.fbIsEqualTo = function cDate_fbIsBefore(oDate) {
    return this.__uYear == oDate.uYear && this.__uMonth == oDate.uMonth && this.__uDay == oDate.uDay;
  };
  cDate.prototype.fbIsAfter = function cDate_fbIsAfter(oDate) {
    if (this.__uYear > oDate.uYear) return true;
    if (this.__uYear < oDate.uYear) return false;
    if (this.__uMonth > oDate.uMonth) return true;
    if (this.__uMonth < oDate.uMonth) return false;
    if (this.__uDay > oDate.uDay) return true;
    //if (this.__uDay < oDate.uDay) return false;
    return false;
  };

  cDate.prototype.fbIsInThePast = function cDate_fbIsInThePast() {
    return this.fbIsBefore(cDate.foNow());
  };
  cDate.prototype.fbIsInThePastUTC = function cDate_fbIsInThePastUTC() {
    return this.fbIsBefore(cDate.foNowUTC());
  };
  cDate.prototype.fbIsToday = function cDate_fbIsToday() {
    return this.fbIsEqualTo(cDate.foNow());
  };
  cDate.prototype.fbIsTodayUTC = function cDate_fbIsTodayUTC() {
    return this.fbIsEqualTo(cDate.foNowUTC());
  };
  cDate.prototype.fbIsInTheFuture = function cDate_fbIsInTheFuture() {
    return this.fbIsAfter(cDate.foNow());
  };
  cDate.prototype.fbIsInTheFutureUTC = function cDate_fbIsInTheFutureUTC() {
    return this.fbIsAfter(cDate.foNowUTC());
  };

  cDate.prototype.fsToHumanReadableString = function cDate_fsToHumanReadableString() {
    // Month <day>th, <year>
    return (
      asMonths[this.__uMonth - 1]
      + " " + this.__uDay.toString() + asOrdinalPostfixes[this.__uDay]
      + ", " + this.__uYear.toString()
    );
  };
  cDate.prototype.foToJSDate = function cDate_toJSDate() {
    return new Date(this.fsToString());
  };
  cDate.prototype.foToJSDateUTC = function cDate_toJSDateUTC() {
    return new Date(this.fsToString() + "T0:0Z");
  };
  cDate.prototype.fxToJSON = function cDate_fxToJSON() {
    return this.fsToString();
  };
  cDate.prototype.fsToString = function cDate_fsToString() {
    return fsGetDateString(this.__uYear, this.__uMonth, this.__uDay);
  };
  cDate.prototype.toString = function cDate_toString() {
    return this.fsToString();
  };
  cDate.prototype.valueOf = function cDate_valueOf() {
    return this.foToJSDate().valueOf();
  };
  return cDate;
})();
